irc_nicktrack.go

```diff
@@ -0,0 +1,103 @@
+package irc
+
+import (
+	"regexp"
+	"strings"
+)
+
+//Struct to store Channel Info
+type Channel struct {
+	Topic string
+	Mode  string
+	Users map[string]User
+}
+
+type User struct {
+	Host string
+	Mode string
+}
+
+var mode_split = regexp.MustCompile("([%@+]{0,1})(.+)") //Half-Op, //Op, //Voice
+
+func (irc *Connection) SetupNickTrack() {
+	// 353: RPL_NAMEREPLY per RFC1459
+	// will typically receive this on channel joins and when NAMES is
+	// called via GetNicksOnCHan
+	irc.AddCallback("353", func(e *Event) {
+		// get chan
+		channelName := e.Arguments[2]
+		// check if chan exists in map
+		_, ok := irc.Channels[channelName]
+
+		// if not make one
+		if ok != true {
+			irc.Channels[channelName] = Channel{Users: make(map[string]User)}
+		}
+		// split the datat into a slice
+		for _, modenick := range strings.Split(e.Message(), " ") {
+			nickandmode := mode_split.FindStringSubmatch(modenick)
+			u := User{}
+			if len(nickandmode) == 3 {
+				if nickandmode[1] == "@" {
+					u.Mode = "+o" // Ooof should be mode struct?
+				} else if nickandmode[1] == "+" {
+					u.Mode = "+v" // Ooof should be mode struct?
+				} else if nickandmode[1] == "%" {
+					u.Mode = "+h"
+				}
+				irc.Channels[channelName].Users[nickandmode[2]] = u
+			} else {
+				irc.Channels[channelName].Users[modenick] = u
+			}
+		}
+	})
+
+	irc.AddCallback("MODE", func(e *Event) {
+		channelName := e.Arguments[0]
+		if len(e.Arguments) == 3 { // 3 == for channel 2 == for user on server
+			if _, ok := irc.Channels[channelName]; ok != true {
+				irc.Channels[channelName] = Channel{Users: make(map[string]User)}
+			}
+			if _, ok := irc.Channels[channelName].Users[e.Arguments[2]]; ok != true {
```

Comment:
This only seems to handle setting all modes, however `MODE` responses don't always contain all modes of a user. Because this code seems to set mode directly, it looks like the Mode won't be accurate if `MODE` is called more than once for a user or if a user has a mode set and `MODE` is called.

Additionally, this will probably end up being the most complicated portion. Either IRCv3 multi-prefix needs to be enabled (and handled) or you will need to query for the user's modes in a channel using `NAMES` or `WHO` (rather, `WHOX`) when a mode is removed. As an example, a user can have `+v+o` in a channel, but in most cases, only the highest mode will be returned (in this case `+o`). When `+o` is removed, they still have `+v`, so there should be a way to determine that.